src/jdk.zipfs/share/classes/jdk/nio/zipfs/ZipFileSystem.java

```diff
@@ -3551,7 +3559,9 @@ public GroupPrincipal group() {
 
         @Override
         public Set<PosixFilePermission> permissions() {
-            return storedPermissions().orElse(Set.copyOf(defaultPermissions));
+            // supportPosix ==> (defaultPermissions != null).
+            assert defaultPermissions != null;
+            return storedPermissions().orElse(defaultPermissions.clone());
```

Comment:
```suggestion
            return storedPermissions().orElseGet(defaultPermissions::clone);
```
So we don't clone if stored permissions is present.

Reply:
*doh* Excellent spot, thanks!